### ISSUE TITLE
multiaddr string representation: Switch to /p2p/QmFoo by default

### ISSUE THREAD

**Issue by Stebalien** [libp2p/libp2p#79]
Proposal: Switch to /p2p/QmFoo multiaddrs (from /ipfs/QmFoo) by default.
Why: The sooner we pull the trigger, the less confusion we'll have to deal with in the long run. js has had support for 5 months, go has had support for 13 months, ~rust has never supported the `/ipfs` syntax~ (see https://github.com/libp2p/libp2p/issues/79#issuecomment-514054461).
Impact: Users of old versions won't be able to parse these new _textual_ multiaddrs. This _does not_ affect network interoperability. This _does not_ prevent new nodes from parsing old addresses.

* [x] Requesting sign-offs from @raulk and @jacobheun.

---

#Tracking:

* Announcement (https://github.com/libp2p/libp2p/issues/79#issuecomment-514414348)
  * [x] Discuss (discuss.libp2p.io, discuss.ipfs.io)
    * [x] Libp2p announcement: https://discuss.libp2p.io/t/multiaddr-format-change-ipfs-p2p/330
    * [x] Ipfs announcement: https://discuss.ipfs.io/t/multiaddr-format-change-ipfs-p2p/6576
  * [x] IRC (#libp2p, #ipfs, #ipfs-dev)
  * [ ] Newsletter? (at least the IPFS one)
  * [ ] Blog Post
* IPFS
  * [ ] docs
    * [ ] ipfs.io: https://github.com/ipfs/website/pull/319
    * [ ] docs.ipfs.io: TODO
  * [ ] config & repo & migration - the ipfs config should use /p2p addresses for bootstrappers.
    * [ ] go
    * [ ] js
  * [x] code/tests
    * [x] js
    * [x] go
      * [x] go-ipfs: https://github.com/ipfs/go-ipfs/pull/6538
      * [x] go-ipfs-config: https://github.com/ipfs/go-ipfs-config/pull/39
* Libp2p
  * [x] docs
    * [x] Specs: https://github.com/libp2p/specs/pull/237
    * [x] Examples: https://github.com/libp2p/go-libp2p-examples/pull/107
  * [x] code/tests
    * [x] js
      * [x] js-multiaddr - https://github.com/multiformats/js-multiaddr/pull/76
      * [x] js-libp2p
    * [x] go
       * [x] go-multiaddr - https://github.com/multiformats/go-multiaddr/pull/115
       * [x] go-libp2p-core
       * [x] go-libp2p

**Comment by tomaka** [comment]
(fwiw, Rust supports and has always supported `/ipfs`)

**Comment by jacobheun** [comment]
JS released p2p support on December 17 of last year (2018) as a patch release, so all new installs got the change. This was a month before the js-ipfs 0.34 release (January 17, 2019), so all installs of that would also have included the change.

We should be good to make the switch over to making this default. Many of the js-libp2p tests have already switched over to specifying `/p2p`. 

I am 👍 for doing this. 

cc: @alanshaw @vasco-santos

**Comment by raulk** [comment]
+1

We'll need to rollout proper change management / communication here:

- [ ] Fully migrate to `p2p` protocol in components we own. A few instances from quick scanning in our go codebase:
    - https://github.com/multiformats/go-multiaddr/blob/71e227b59a65e7642900cc2cdeecd3773f779598/protocols.go#L138
    - https://github.com/libp2p/go-libp2p-examples/blob/01f8a674bd08eb5156ba46a99893ab72c7668731/ipfs-camp-2019/04-Protocol/main.go#L55
    - https://github.com/libp2p/go-libp2p-gorpc/blob/7cd5ddd4060f94c90ff7bbb0450b84461e64628a/examples/ping/ping.go#L56
- [ ] Removal of functions and logic that supported the transition (inducing breakage), e.g.:
    - https://github.com/multiformats/go-multiaddr/blob/1ae13209557ac03cb1ee628d3f3550b46b2f4f88/protocol.go#L75
    - https://github.com/multiformats/go-multiaddr/blob/1ae13209557ac03cb1ee628d3f3550b46b2f4f88/protocol.go#L65
- [ ] Proper semver version bump in components where this change originates, percolating the same bump up to affected dependents.
- [ ] Make sure release notes warn about this change prominently.
- [ ] Communication in discuss.libp2p.io, discuss.ipfs.io, and in IRC channels.
- [ ] Notice via an issue in top-level repos (for GitHub subscribers), redirecting to the relevant discussions in the boards.
- [ ] Profit ✅

**Comment by raulk** [comment]
(I took the liberty to update the title to clarify this is about the _string representation_ of certain multiaddrs).

**Comment by jacobheun** [comment]
@raulk it sounds like you're talking about removing support for `/ipfs` entirely? Or am I mistaken? We should still support it for some time after switching to `/p2p` by default.

I agree that communication is important for this, as anyone running older versions will not be able to properly communicate with the latest nodes.

For js this will be a simple change of flipping the order of lines here, https://github.com/multiformats/js-multiaddr/blob/v6.1.0/src/protocols-table.js#L45-L50, doing a major version release, and upgrading that in the libp2p (and js-ipfs) repos. I'll create an issue in js-libp2p to track this and update it with any timelines/communication channels we decide.

**Comment by raulk** [comment]
@jacobheun you're right. Re-reading this issue, it proposes a change on the encoding side, such that `p2p` components are serialised as `/p2p/` rather than `/ipfs/`. Indeed, this would only affect old clients that are unable to serialize `/p2p/` tokens.

**Comment by jacobheun** [comment]
Started https://github.com/libp2p/js-libp2p/issues/386 to track this for JS. I will update that with any announcement/dates we decide on. JS will likely just aim for upcoming minor (breaking) releases of js-libp2p and js-ipfs, unless we need/want something more specific.

**Comment by Stebalien** [comment]
@tomaka you _sure_ about that?

**Comment by Stebalien** [comment]
@raulk let's

* Definitely announce this change widely.
* Not remove support for parsing /ipfs addresses. Honestly, we probably never need to remove support.

**Comment by tomaka** [comment]
> @tomaka you _sure_ about that?

I checked after I saw you ask, and we indeed removed support. Sorry for the false info.

**Comment by Stebalien** [comment]
Proposed communication:


---

Libp2p is replacing all multiaddrs of the form `/ipfs/QmFoo` (old-style) with `/p2p/QmFoo` (new-style). The `/ipfs` prefix his an historical artifact from the days before libp2p was split out of IPFS.

All libp2p implementations have supported _parsing_ new-style multiaddrs for at least 7 months (see the [timeline](#timeline)). We are now entering the second phase of this transition and will:

1. Deprecate the old-style (without removing support).
2. Start formatting addresses using the new-style by default.

Impact:

* This change does not affect the network or binary representation of multiaddrs, just the textual representation.
* This change will not affect network interoperability.
* Old nodes (see below) will not be able to parse textual new-style multiaddrs. Specifically, `ipfs swarm connect /p2p/QmFoobar` will not work on go-ipfs releases before 0.4.16 or js-ipfs releases before 0.34.

## Timeline

* go-libp2p - June 26, 2018 (tag gx/v6.0.5)
* go-ipfs - July 13, 2018 (0.4.16)
* rust-libp2p - Since the beginning.
* js-libp2p - December 17, 2018
* js-ipfs - January 17, 2019 (0.34)

**Comment by Stebalien** [comment]
So, it looks like js-libp2p has already switched. Given that go-libp2p is now the only libp2p implementation that uses /ipfs/QmFooBar by default..., I'm going to publish this.
